src/learning/agglomerative.rs

```diff
@@ -0,0 +1,415 @@
+//! Agglomerative (Hierarchical) Clustering Module
+//!
+//! Contains implementation of Agglomerative Clustering.
+//!
+//! # Usage
+//!
+//! ```
+//! use rusty_machine::learning::agglomerative::{AgglomerativeClustering, Metrics};
+//! use rusty_machine::learning::SupModel;
+//! use rusty_machine::linalg::{Matrix, Vector};
+//!
+//! let inputs = Matrix::new(4, 2, vec![1., 3., 2., 3., 4., 3., 5., 3.]);
+//! let mut agg = AgglomerativeClustering::new(2, Metrics::Single);
+//!
+//! // Train the model and get the clustering result
+//! let res = agg.train(&inputs).unwrap();
+//!
+//! assert_eq!(res, Vector::new(vec![0, 0, 1, 1]));
+//! ```
+//! ```
+
+use std::collections::{BTreeMap, HashMap};
+use std::f64;
+
+use linalg::{Matrix, BaseMatrix, Vector};
+use learning::{LearningResult};
+
+/// Agglomerative clustering distances
+#[derive(Debug)]
+pub enum Metrics {
```

Comment:
I'm not sure if this is relevant here but I would consider using a trait here instead of an enum. The upside is that users can define their own metrics and plug them into the model (without modifying rusty-machine source code). The downside is an extra generic parameter on the model.